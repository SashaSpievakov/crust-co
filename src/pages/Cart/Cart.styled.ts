import styled from 'styled-components';
import { BsCart3, BsTrash, BsChevronLeft } from 'react-icons/bs';
import { Container } from 'src/styles/Base.styled';

const Article = styled(Container)`
  margin: 50px auto;

  @media screen and (max-width: 400px) {
    margin: 30px auto;
  }
`;

const Top = styled.section`
  display: flex;
  justify-content: space-between;
  align-items: center;

  @media screen and (max-width: 400px) {
    flex-direction: column;
    gap: 15px;
  }
`;

const Title = styled.h2`
  display: flex;
  align-items: center;
  font-size: 32px;
`;

const CartIcon = styled(BsCart3)`
  stroke-width: 0.5;
  margin-right: 10px;
`;

const Clear = styled.section`
  display: flex;
  align-items: center;
  border-radius: 5px;
  cursor: pointer;

  &:focus-visible {
    background-color: ${({ theme }) => theme.colors.gray};
    outline: 5px solid ${({ theme }) => theme.colors.gray};
    outline-offset: -1px;
  }

  span {
    display: inline-block;
    margin-left: 7px;
    font-size: 18px;
  }

  &:hover {
    color: ${({ theme }) => theme.colors.secondary};

    span {
      color: ${({ theme }) => theme.colors.secondary};
    }
  }
`;

const Trash = styled(BsTrash)`
  font-size: 18px;
  stroke-width: 0.3;
`;

const Bottom = styled.section`
  margin-top: 30px;
  border-top: 1px solid ${({ theme }) => theme.colors.gray};
`;

const Details = styled.section`
  font-size: 22px;
  display: flex;
  justify-content: space-between;
  margin: 40px 50px 0;

  @media screen and (max-width: 400px) {
    margin: 40px 25px 0;
  }
`;

const Buttons = styled.section`
  display: flex;
  justify-content: space-between;
  margin-top: 40px;

  @media screen and (max-width: 600px) {
    flex-direction: column-reverse;
    align-items: center;
    gap: 20px;
  }
`;

const ChevronLeft = styled(BsChevronLeft)`
  stroke-width: 1.3;
  margin-right: 12px;
`;

export {
  Article,
  Top,
  Title,
  CartIcon,
  Clear,
  Trash,
  Bottom,
  Details,
  Buttons,
  ChevronLeft,
};
