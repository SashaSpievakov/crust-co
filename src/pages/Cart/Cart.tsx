import { Link } from 'react-router-dom';
import { shallowEqual } from 'react-redux';

import { clearCart } from '../../store/slices/cart/reducer/cartReducer';
import { selectCart } from '../../store/slices/cart/selectors/selectCart/selectCart';
import CartItem from '../../components/CartItem/CartItem';
import CartEmpty from '../../components/CartEmpty/CartEmpty';
import { ButtonBackCart, ButtonPay } from '../../styles/Buttons.styled';
import {
  Article,
  Top,
  Title,
  CartIcon,
  Clear,
  Trash,
  Bottom,
  Details,
  Buttons,
  ChevronLeft,
} from './Cart.styled';
import { ICartItem } from '../../models/ICartItem';
import { useAppDispatch, useAppSelector } from '../../hooks/reduxHooks';

const Cart = () => {
  const { itemsCount, totalPrice, items } = useAppSelector(
    selectCart,
    shallowEqual,
  );
  const dispatch = useAppDispatch();

  const onClearClick = () => {
    dispatch(clearCart());
  };

  return items.length ? (
    <Article>
      <Top>
        <Title>
          <CartIcon />
          Cart
        </Title>
        <Clear onClick={onClearClick}>
          <Trash />
          <span role="presentation">Delete all items</span>
        </Clear>
      </Top>

      <section>
        {items.map((item: ICartItem) => (
          <CartItem key={item.name + item.type + item.size} {...item} />
        ))}
      </section>

      <Bottom>
        <Details>
          <span>
            {' '}
            Items <b data-testid="cartPageItemsCount">{itemsCount}</b>{' '}
          </span>
          <span>
            {' '}
            Total <b data-testid="cartPageItemsPrice">{totalPrice}$</b>{' '}
          </span>
        </Details>
        <Buttons>
          <ButtonBackCart to="/" as={Link}>
            <ChevronLeft />
            <span>Go back</span>
          </ButtonBackCart>
          <ButtonPay>
            <span>Buy now</span>
          </ButtonPay>
        </Buttons>
      </Bottom>
    </Article>
  ) : (
    <CartEmpty />
  );
};

export default Cart;
