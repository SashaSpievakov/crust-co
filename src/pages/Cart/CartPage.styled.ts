import { BsCart3, BsChevronLeft, BsTrash } from 'react-icons/bs';
import { Container } from 'src/styles/Base.styled';
import styled from 'styled-components';

const Article = styled(Container)`
  margin: 50px auto;

  @media screen and (max-width: 400px) {
    margin: 30px auto;
  }
`;

const Top = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;

  @media screen and (max-width: 400px) {
    flex-direction: column;
    gap: 15px;
  }
`;

const Title = styled.h2`
  display: flex;
  align-items: center;
  font-size: 32px;
`;

const CartIcon = styled(BsCart3)`
  stroke-width: 0.5;
  margin-right: 10px;
`;

const Clear = styled.section`
  display: flex;
  align-items: center;
  border-radius: 5px;
  cursor: pointer;

  &:focus-visible {
    background-color: ${({ theme }) => theme.colors.gray};
    outline: 5px solid ${({ theme }) => theme.colors.gray};
    outline-offset: -1px;
  }

  &:hover {
    color: ${({ theme }) => theme.colors.secondary};
  }
`;

const Trash = styled(BsTrash)`
  font-size: 18px;
  stroke-width: 0.3;
`;

const DeleteText = styled.span`
  display: inline-block;
  margin-left: 7px;
  font-size: 18px;

  &:hover {
    color: ${({ theme }) => theme.colors.secondary};
  }
`;

const Bottom = styled.section`
  margin-top: 30px;
  border-top: 1px solid ${({ theme }) => theme.colors.gray};
`;

const Details = styled.section`
  font-size: 22px;
  display: flex;
  justify-content: space-between;
  margin: 40px 50px 0;

  @media screen and (max-width: 400px) {
    margin: 40px 25px 0;
  }
`;

const Buttons = styled.section`
  display: flex;
  justify-content: space-between;
  margin-top: 40px;

  @media screen and (max-width: 600px) {
    flex-direction: column-reverse;
    align-items: center;
    gap: 20px;
  }
`;

const ChevronLeft = styled(BsChevronLeft)`
  stroke-width: 1.3;
  margin-right: 12px;
`;

export {
  Article,
  Bottom,
  Buttons,
  CartIcon,
  ChevronLeft,
  Clear,
  DeleteText,
  Details,
  Title,
  Top,
  Trash,
};
