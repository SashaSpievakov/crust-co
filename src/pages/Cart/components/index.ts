export { CartEmpty } from './CartEmpty';
export { CartItem } from './CartItem';
export { OrderModal } from './OrderModal';
