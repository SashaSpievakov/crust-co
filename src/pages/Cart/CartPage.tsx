import { FC, KeyboardEvent, useEffect, useState } from 'react';
import { shallowEqual } from 'react-redux';
import { Link } from 'react-router-dom';

import { useAppDispatch, useAppSelector } from '../../hooks';
import { ICartItem } from '../../models';
import { clearCart } from '../../store';
import { selectCart } from '../../store/slices/cart/selectors/selectCart/selectCart';
import { ButtonBackCart, ButtonPay } from '../../styles/Buttons.styled';
import {
  Article,
  Bottom,
  Buttons,
  CartIcon,
  ChevronLeft,
  Clear,
  DeleteText,
  Details,
  Title,
  Top,
  Trash,
} from './CartPage.styled';
import { CartEmpty, CartItem, OrderModal } from './components';

export const CartPage: FC = () => {
  const [isOpen, setIsOpen] = useState<boolean>(false);
  const { itemsCount, totalPrice, items } = useAppSelector(
    selectCart,
    shallowEqual,
  );
  const dispatch = useAppDispatch();

  const onClearClick = () => {
    dispatch(clearCart());
  };

  const handleClearKeyDownClick = (e: KeyboardEvent<HTMLElement>) => {
    if (e.code === 'Enter') onClearClick();
  };

  useEffect(() => {
    document.title = 'Crust & Co. | Cart';
  }, []);

  return items.length ? (
    <Article>
      <Top>
        <Title>
          <CartIcon />
          Cart
        </Title>
        <Clear
          onClick={onClearClick}
          onKeyDown={(e) => handleClearKeyDownClick(e)}
          tabIndex={0}
        >
          <Trash />
          <DeleteText role="presentation">Delete all items</DeleteText>
        </Clear>
      </Top>

      <section>
        {items.map((item: ICartItem) => (
          <CartItem key={item.name + item.type + item.size} {...item} />
        ))}
      </section>

      <Bottom>
        <Details>
          <span>
            {' '}
            Items <b data-testid="cartPageItemsCount">{itemsCount}</b>{' '}
          </span>
          <span>
            {' '}
            Total <b data-testid="cartPageItemsPrice">{totalPrice}$</b>{' '}
          </span>
        </Details>
        <Buttons>
          <ButtonBackCart to="/" as={Link}>
            <ChevronLeft />
            <span>Go back</span>
          </ButtonBackCart>
          <ButtonPay onClick={() => setIsOpen(true)}>
            <span>Buy now</span>
          </ButtonPay>
          {isOpen && <OrderModal setIsOpen={setIsOpen} />}
        </Buttons>
      </Bottom>
    </Article>
  ) : (
    <CartEmpty />
  );
};
