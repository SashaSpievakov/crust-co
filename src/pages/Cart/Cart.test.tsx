import '@testing-library/jest-dom';

import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import {
  rendererWithProviders,
  renderWithProvidersAndRoutes,
} from '../../tests/helpers';
import { CartItemsMockProps } from '../../tests/mocks/mockData/mockData';
import Cart from './Cart';

describe('Cart Tests', () => {
  test('renders the Cart page', () => {
    const snapshot = rendererWithProviders(<Cart />);
    expect(snapshot).toMatchSnapshot();
  });

  test('renders the Cart page with items', () => {
    const snapshot = rendererWithProviders(<Cart />, '/', {
      cart: CartItemsMockProps,
    });
    expect(snapshot).toMatchSnapshot();
  });

  test('deletes all items from the cart', () => {
    renderWithProvidersAndRoutes(<Cart />, false, '/', {
      cart: CartItemsMockProps,
    });
    const deleteButton = screen.getByText(/delete all items/i);

    userEvent.click(deleteButton);

    expect(screen.getByRole('heading')).toHaveTextContent(
      /your cart is empty ðŸ˜•/i,
    );
  });

  test('checks redirect to the Home page', () => {
    renderWithProvidersAndRoutes(null, true, '/cart', {
      cart: CartItemsMockProps,
    });
    const link = screen.getByRole('link', {
      name: /go back/i,
    });

    userEvent.click(link);

    expect(screen.getByText(/all pizzas/i)).toBeInTheDocument();
  });
});
