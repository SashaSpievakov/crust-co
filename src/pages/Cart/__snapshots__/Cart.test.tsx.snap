// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart Tests renders the Cart page 1`] = `
Object {
  "children": Array [
    <h2>
      Your cart is empty 
      <span>
        ðŸ˜•
      </span>
    </h2>,
    <p>
      You haven't ordered food yet.
      <br />
      If you want to make an order, go back to the main page.
    </p>,
    <img
      alt="empty cart"
      height="255"
      src="empty-cart.png"
      width="300"
    />,
    <a
      className="sc-hLBbgP sc-eDvSVe fGqcTx lhIsum"
      href="/"
      onClick={[Function]}
    >
      Go Back
    </a>,
  ],
  "props": Object {
    "className": "sc-gikAfH jgwSHN",
  },
  "type": "article",
}
`;

exports[`Cart Tests renders the Cart page with items 1`] = `
Object {
  "children": Array [
    <section
      className="sc-cjibBx eMsni"
    >
      <h2
        className="sc-jcMfQk kyZZBh"
      >
        <svg
          className="sc-cabOPr hrQlNv"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 1.5A.5.5 0 0 1 .5 1H2a.5.5 0 0 1 .485.379L2.89 3H14.5a.5.5 0 0 1 .49.598l-1 5a.5.5 0 0 1-.465.401l-9.397.472L4.415 11H13a.5.5 0 0 1 0 1H4a.5.5 0 0 1-.491-.408L2.01 3.607 1.61 2H.5a.5.5 0 0 1-.5-.5zM3.102 4l.84 4.479 9.144-.459L13.89 4H3.102zM5 12a2 2 0 1 0 0 4 2 2 0 0 0 0-4zm7 0a2 2 0 1 0 0 4 2 2 0 0 0 0-4zm-7 1a1 1 0 1 1 0 2 1 1 0 0 1 0-2zm7 0a1 1 0 1 1 0 2 1 1 0 0 1 0-2z"
          />
        </svg>
        Cart
      </h2>
      <section
        className="sc-iTFTee bDKgHL"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={0}
      >
        <svg
          className="sc-iAEawV fqfnEp"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5.5 5.5A.5.5 0 0 1 6 6v6a.5.5 0 0 1-1 0V6a.5.5 0 0 1 .5-.5zm2.5 0a.5.5 0 0 1 .5.5v6a.5.5 0 0 1-1 0V6a.5.5 0 0 1 .5-.5zm3 .5a.5.5 0 0 0-1 0v6a.5.5 0 0 0 1 0V6z"
          />
          <path
            d="M14.5 3a1 1 0 0 1-1 1H13v9a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V4h-.5a1 1 0 0 1-1-1V2a1 1 0 0 1 1-1H6a1 1 0 0 1 1-1h2a1 1 0 0 1 1 1h3.5a1 1 0 0 1 1 1v1zM4.118 4 4 4.059V13a1 1 0 0 0 1 1h6a1 1 0 0 0 1-1V4.059L11.882 4H4.118zM2.5 3V2h11v1h-11z"
            fillRule="evenodd"
          />
        </svg>
        <span
          role="presentation"
        >
          Delete all items
        </span>
      </section>
    </section>,
    <section>
      <article
        className="sc-ezOQGI bqqjuW"
      >
        <div
          className="sc-bYMpWt bnFYVZ"
        >
          <picture>
            <source
              srcSet="./assets/img/pizza6.webp"
              type="image/webp"
            />
            <source
              srcSet="./assets/img/pizza6.png"
              type="image/png"
            />
            <img
              alt="Cheesburger Pizza pizza"
              height="80"
              src="./assets/img/pizza6.png"
              width="80"
            />
          </picture>
        </div>
        <section
          className="sc-kMjNwy gJsxfE"
        >
          <h3>
            Cheesburger Pizza
          </h3>
          <p>
            thin
             dough, 
            14
             inch
          </p>
        </section>
        <section
          className="sc-ilhmMj fPuPTA"
        >
          <button
            aria-label="minus"
            className="sc-hLBbgP sc-iBYQkv sc-ftTHYK fGqcTx iDltE YugEl"
            disabled={false}
            onClick={[Function]}
          >
            <svg
              className="sc-jIRcFI kihpWA"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 8a1 1 0 0 1 1-1h14a1 1 0 1 1 0 2H1a1 1 0 0 1-1-1z"
              />
            </svg>
          </button>
          <span>
            13
          </span>
          <button
            aria-label="plus"
            className="sc-hLBbgP sc-iBYQkv sc-ftTHYK fGqcTx iDltE YugEl"
            onClick={[Function]}
          >
            <svg
              className="sc-jIRcFI kihpWA"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 0a1 1 0 0 1 1 1v6h6a1 1 0 1 1 0 2H9v6a1 1 0 1 1-2 0V9H1a1 1 0 0 1 0-2h6V1a1 1 0 0 1 1-1z"
              />
            </svg>
          </button>
        </section>
        <h3
          className="sc-hhOBVt kgouOf"
        >
          91
          $
        </h3>
        <div
          className="sc-gYbzsP kNflxz"
        >
          <button
            aria-label="delete"
            className="sc-hLBbgP sc-iBYQkv sc-ftTHYK fGqcTx iDltE kTqzjG"
            onClick={[Function]}
          >
            <svg
              className="sc-jIRcFI kihpWA"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 0a1 1 0 0 1 1 1v6h6a1 1 0 1 1 0 2H9v6a1 1 0 1 1-2 0V9H1a1 1 0 0 1 0-2h6V1a1 1 0 0 1 1-1z"
              />
            </svg>
          </button>
        </div>
      </article>
      <article
        className="sc-ezOQGI bqqjuW"
      >
        <div
          className="sc-bYMpWt bnFYVZ"
        >
          <picture>
            <source
              srcSet="./assets/img/pizza2.webp"
              type="image/webp"
            />
            <source
              srcSet="./assets/img/pizza2.png"
              type="image/png"
            />
            <img
              alt="Pepperoni Pizza pizza"
              height="80"
              src="./assets/img/pizza2.png"
              width="80"
            />
          </picture>
        </div>
        <section
          className="sc-kMjNwy gJsxfE"
        >
          <h3>
            Pepperoni Pizza
          </h3>
          <p>
            traditional
             dough, 
            16
             inch
          </p>
        </section>
        <section
          className="sc-ilhmMj fPuPTA"
        >
          <button
            aria-label="minus"
            className="sc-hLBbgP sc-iBYQkv sc-ftTHYK fGqcTx iDltE YugEl"
            disabled={true}
            onClick={[Function]}
          >
            <svg
              className="sc-jIRcFI kihpWA"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 8a1 1 0 0 1 1-1h14a1 1 0 1 1 0 2H1a1 1 0 0 1-1-1z"
              />
            </svg>
          </button>
          <span>
            1
          </span>
          <button
            aria-label="plus"
            className="sc-hLBbgP sc-iBYQkv sc-ftTHYK fGqcTx iDltE YugEl"
            onClick={[Function]}
          >
            <svg
              className="sc-jIRcFI kihpWA"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 0a1 1 0 0 1 1 1v6h6a1 1 0 1 1 0 2H9v6a1 1 0 1 1-2 0V9H1a1 1 0 0 1 0-2h6V1a1 1 0 0 1 1-1z"
              />
            </svg>
          </button>
        </section>
        <h3
          className="sc-hhOBVt kgouOf"
        >
          9
          $
        </h3>
        <div
          className="sc-gYbzsP kNflxz"
        >
          <button
            aria-label="delete"
            className="sc-hLBbgP sc-iBYQkv sc-ftTHYK fGqcTx iDltE kTqzjG"
            onClick={[Function]}
          >
            <svg
              className="sc-jIRcFI kihpWA"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 0a1 1 0 0 1 1 1v6h6a1 1 0 1 1 0 2H9v6a1 1 0 1 1-2 0V9H1a1 1 0 0 1 0-2h6V1a1 1 0 0 1 1-1z"
              />
            </svg>
          </button>
        </div>
      </article>
    </section>,
    <section
      className="sc-eeMvmM gZuIyJ"
    >
      <section
        className="sc-cUEOzv jEoHkZ"
      >
        <span>
           
          Items 
          <b
            data-testid="cartPageItemsCount"
          >
            14
          </b>
           
        </span>
        <span>
           
          Total 
          <b
            data-testid="cartPageItemsPrice"
          >
            100
            $
          </b>
           
        </span>
      </section>
      <section
        className="sc-jOiSOi hJIfdV"
      >
        <a
          className="sc-hLBbgP sc-iBYQkv sc-pyfCe sc-jrcTuL fGqcTx iDltE uzlJh hoiVoq"
          href="/"
          onClick={[Function]}
        >
          <svg
            className="sc-dPWrhe bcZsET"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11.354 1.646a.5.5 0 0 1 0 .708L5.707 8l5.647 5.646a.5.5 0 0 1-.708.708l-6-6a.5.5 0 0 1 0-.708l6-6a.5.5 0 0 1 .708 0z"
              fillRule="evenodd"
            />
          </svg>
          <span>
            Go back
          </span>
        </a>
        <button
          className="sc-hLBbgP sc-gKPRtg fGqcTx fdWAyy"
          onClick={[Function]}
        >
          <span>
            Buy now
          </span>
        </button>
      </section>
    </section>,
  ],
  "props": Object {
    "className": "sc-bcXHqe sc-cCjUiG bZXhLi bvyhPU",
  },
  "type": "article",
}
`;
